securesystemslib/signer.py

```diff
@@ -137,8 +150,264 @@ def to_dict(self) -> Dict:
         }
 
 
+class Key:
+    """Abstract class representing the public portion of a key.
+
+    *All parameters named below are not just constructor arguments but also
+    instance attributes.*
+
+    Args:
+        keyid: Key identifier that is unique within the metadata it is used in.
+            Keyid is not verified to be the hash of a specific representation
+            of the key.
+        keytype: Key type, e.g. "rsa", "ed25519" or "ecdsa-sha2-nistp256".
+        scheme: Signature scheme. For example:
+            "rsassa-pss-sha256", "ed25519", and "ecdsa-sha2-nistp256".
+        keyval: Opaque key content
+        unrecognized_fields: Dictionary of all attributes that are not managed
+            by Securesystemslib
+
+    Raises:
+        TypeError: Invalid type for an argument.
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(
+        self,
+        keyid: str,
+        keytype: str,
+        scheme: str,
+        keyval: Dict[str, Any],
```

Comment:
`Key` really only needs `keytype` and `scheme` to dispatch. We could define all other fields in sublcasses for more flexibility.  `SSlibKey` can still be the default to ensure spec compliance.

Reply:
I don't understand what we gain with that -- sure, Key itself doesn't use them but every Key implementattion that is compatible with intoto or TUF spec does need them.  If Key is an implementation of TUF and in-toto specification keys, why would we remove fields defined in those specs? It just seems like every spec-compliant key has to re-implement them anyway.

Reply:
> I don't understand what we gain with that

We would gain a base class that can be used in a meaningful way for non-spec compliant keys. As `GPGKey` shows this is not an unlikely use case.

I understand that you want to avoid to accidentally load a non-spec compliant key with `Key.from_dict`, but I think that this can also be achieved by making `SSlibKey` the strictly spec-conformant key, and registering only that in the default `KEY_FOR_TYPE_AND_SCHEME`. Moreover, python-tuf could require `SSlibKey` instances.

Reply:
Btw. using `scheme` alone to dispatch, as I proposed below, would allow us even remove the `keytype` field from `Key`.

Reply:
I won't insist on either of them. I just think that TUF and in-toto public key metadata formats are not ideal and that we can design the `Key` interface to make it easier for implementations that use a better key format (at the cost of being non-spec compliant).

Reply:
> I understand that you want to avoid to accidentally load a non-spec compliant key with Key.from_dict, but I think that this can also be achieved by making SSlibKey the strictly spec-conformant key, and registering only that in the default KEY_FOR_TYPE_AND_SCHEME.

Oh! I don't think this is something I want: I'd certainly want to add things like "SigstoreKey" and "CompatibleGPGKey" in the future (not necessarily in securesystemslib but somewhere), and I would not want to put them in SSlibKey